kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -0,0 +1,162 @@
+
+      // Process all checkpoint batches
+      while (checkpointIterator.hasNext()) {
+        FilteredColumnarBatch filteredBatch = checkpointIterator.next();
+        ColumnarBatch batch = filteredBatch.getData();
+        Optional<ColumnVector> selectionVector = filteredBatch.getSelectionVector();
+        final int rowCount = batch.getSize();
+
+        ColumnVector removeVector = batch.getColumnVector(REMOVE_INDEX);
+        ColumnVector addVector = batch.getColumnVector(ADD_INDEX);
+        ColumnVector domainMetadataVector = batch.getColumnVector(DOMAIN_METADATA_INDEX);
+
+        // Process all selected rows in a single pass for optimal performance
+        for (int i = 0; i < rowCount; i++) {
+          // Check if this row is selected
+          boolean isSelected =
+              !selectionVector.isPresent()
+                  || (!selectionVector.get().isNullAt(i) && selectionVector.get().getBoolean(i));
+          if (!isSelected) continue;
+
+          // Step 1: Ensure there are no remove records
+          // We set minFileRetentionTimestampMillis to infinite future to skip all removed files,
+          // so there should be no remove actions.
+          checkState(
+              removeVector.isNullAt(i),
+              "unexpected remove row found when "
+                  + "setting minFileRetentionTimestampMillis to infinite future");
+
+          // Step 2: Process add file records - direct columnar access for better performance
+          if (!addVector.isNullAt(i)) {
+            // Direct access to size field in add file record for maximum performance
+            ColumnVector sizeVector = addVector.getChild(ADD_SIZE_INDEX);
+            long fileSize = sizeVector.getLong(i);
+
+            tableSizeByte.add(fileSize);
+            fileSizeHistogram.insert(fileSize);
+            fileCount.increment();
+          }
+
+          // Step 3: Process domain metadata records
+          if (!domainMetadataVector.isNullAt(i)) {
+            DomainMetadata domainMetadata =
+                DomainMetadata.fromColumnVector(domainMetadataVector, i);
+            if (domainMetadata != null
+                && !domainMetadataMap.containsKey(domainMetadata.getDomain())) {
+              domainMetadataMap.put(domainMetadata.getDomain(), domainMetadata);
+            }
+          }
+        }
+      }
+
+      // Write the checksum file
+      try {
+        checksumWriter.writeCheckSum(
+            engine,
+            new CRCInfo(
+                snapshot.getVersion(),
+                snapshot.getMetadata(),
+                snapshot.getProtocol(),
+                tableSizeByte.longValue(),
+                fileCount.longValue(),
+                Optional.empty() /* txnId */,
+                Optional.of(
+                    domainMetadataMap.values().stream()
+                        .filter(domainMetadata -> !domainMetadata.isRemoved())
+                        .collect(Collectors.toSet())),
```

Comment:
Can't we do this filtering during the iteration? I.e. only save the DM if it's not removed?